pkg/spade-test/spade-test.go

```diff
@@ -0,0 +1,286 @@
+			logger.Info(err)
+			errChan <- err
+		} else {
+			resultChan <- blk
+		}
+	}()
+	select {
+	case <-notFound:
+		return nil, errors.New("DONT_HAVE received from the target peer")
+
+	case blk := <-resultChan:
+		return blk, nil
+
+	case err := <-errChan:
+		return nil, fmt.Errorf("error received %s", err)
+	}
+}
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes
+// Returns true if all retrievals were successful, false if any failed
+func SpadeTraversal(ctx context.Context, startingCid goCid.Cid, p peer.AddrInfo) (bool, error) {
+	cidToRetrieve := startingCid
+
+	// It will always be 3 layers, no more no less
+	// ref: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
+	for i := 0; i <= 3; i++ {
+		// For some reason, need to re-init the host & client every time we do a fetch
+		// otherwise, we get context timeout error after the first fetch
+		host, err := net.InitHost(ctx, nil)
+		if err != nil {
+			return false, fmt.Errorf("failed to init host %s", err)
+		}
+
+		client := NewBitswapClient(host, time.Second*1)
+
+		// Retrieval
+		logger.Debugf("retrieving %s\n", cidToRetrieve.String())
+		blk, err := client.RetrieveBlock(ctx, p, cidToRetrieve)
+		if err != nil {
+			return false, fmt.Errorf("unable to retrieve cid %s", err)
+		}
+
+		if i == 3 {
+			// we've reached the bottom of the tree
+			logger.Debugf("retrieved data cid %s which contains %d bytes\n", cidToRetrieve.String(), len(blk.RawData()))
+			return true, nil
+		}
+
+		// if not at bottom of the tree, keep going down the links until we reach it
+		links, err := FindLinks(ctx, blk)
+		if err != nil {
+			log.Fatalf("unable to find links %s", err)
+		}
+
+		logger.Debugf("retrieved %s which has %d links\n", cidToRetrieve.String(), len(links))
+
+		nextIndex := 0
+		rand.Seed(time.Now().UnixNano())
+		if i == 0 {
+			if len(links) == 1 {
+				return false, fmt.Errorf("starting node only contains one link which must be the manifest")
+			}
+
+			// from the starting node's children, never grab the first link as it refers to the AggregateManifest
+			nextIndex = 1 + rand.Intn(len(links)-1)
+		} else {
+			// randomly pick a link to go down
+			nextIndex = rand.Intn(len(links))
+		}
+
+		cidToRetrieve, err = cid.Parse(links[nextIndex].String())
+		if err != nil {
+			return false, fmt.Errorf("unable to parse cid %s", err)
+		}
+	}
+
+	return false, nil
+}
+
+func main() {
+	ctx := context.Background()
+	logging.SetLogLevel("spade-test", "DEBUG")
+	logger.Debugf("starting spade-test")
+
+	cidToRetrieve, err := cid.Parse("bafybeib62b4ukyzjcj7d2h4mbzjgg7l6qiz3ma4vb4b2bawmcauf5afvua")
+	if err != nil {
+		log.Fatalf("unable to parse cid %s", err)
+	}
+
+	peerID, err := peer.Decode("12D3KooWNrzJ4aeavdsuxkGpErb33G7Daf2FmX8bJHx9bdE6WFzG")
+	if err != nil {
+		log.Fatalf("unable to decode peerID %s", err)
+	}
+
+	addrs, err := convert.StringArrayToMultiaddrs([]string{"/ip4/127.0.0.1/tcp/4001"})
```

Comment:
test fixture stuff - this is just making it work with my locally running ipfs daemon. prod one will take the sp addrInfo from the task queue